// @copilotkit/react-ui/dist/chunk-6U3O2JZP.mjs
import {
  Messages
} from "./chunk-73EBDGYK.mjs";
import {
  ResponseButton
} from "./chunk-YEHO5VMA.mjs";
import {
  Window
} from "./chunk-B7DNOYVQ.mjs";
import {
  Button
} from "./chunk-YTIGBBTC.mjs";
import {
  Header
} from "./chunk-7JYUCW7H.mjs";
import {
  Input
} from "./chunk-KE3N45ZY.mjs";
import {
  ChatContextProvider
} from "./chunk-7LMXXGJT.mjs";
import {
  __async
} from "./chunk-MRXNTQOX.mjs";

// src/components/chat/Chat.tsx
import React, { useEffect } from "react";
import { useCopilotChat } from "@copilotkit/react-core";
import { nanoid } from "nanoid";
import { jsx, jsxs } from "react/jsx-runtime";
var CopilotChat = ({
  instructions,
  defaultOpen = false,
  clickOutsideToClose = true,
  hitEscapeToClose = true,
  onSetOpen,
  onSubmitMessage,
  shortcut = "/",
  icons,
  labels,
  makeSystemMessage,
  showResponseButton = true,
  onInProgress,
  Window: Window2 = Window,
  Button: Button2 = Button,
  Header: Header2 = Header,
  Messages: Messages2 = Messages,
  Input: Input2 = Input,
  ResponseButton: ResponseButton2 = ResponseButton,
  className,
  children
}) => {
  const { visibleMessages, append, reload, stop, isLoading, input, setInput } = useCopilotChat({
    id: nanoid(),
    makeSystemMessage,
    additionalInstructions: instructions
  });
  useEffect(() => {
    onInProgress == null ? void 0 : onInProgress(isLoading);
  }, [isLoading]);
  const [openState, setOpenState] = React.useState(defaultOpen);
  const setOpen = (open) => {
    onSetOpen == null ? void 0 : onSetOpen(open);
    setOpenState(open);
  };
  const sendMessage = (message) => __async(void 0, null, function* () {
    onSubmitMessage == null ? void 0 : onSubmitMessage(message);
    append({
      id: nanoid(),
      content: message,
      role: "user"
    });
  });
  return /* @__PURE__ */ jsxs(ChatContextProvider, { icons, labels, open: openState, setOpen: setOpenState, children: [
    children,
    /* @__PURE__ */ jsxs("div", { className, children: [
      /* @__PURE__ */ jsx(Button2, { open: openState, setOpen }),
      /* @__PURE__ */ jsxs(
        Window2,
        {
          open: openState,
          setOpen,
          clickOutsideToClose,
          shortcut,
          hitEscapeToClose,
          children: [
            /* @__PURE__ */ jsx(Header2, { open: openState, setOpen }),
            /* @__PURE__ */ jsx(Messages2, { messages: visibleMessages, inProgress: isLoading }),
            /* @__PURE__ */ jsx(Input2, { inProgress: isLoading, onSend: sendMessage, isVisible: openState, children: showResponseButton && visibleMessages.length > 0 && /* @__PURE__ */ jsx(ResponseButton2, { onClick: isLoading ? stop : reload, inProgress: isLoading }) })
          ]
        }
      )
    ] })
  ] });
};

export {
  CopilotChat
};
